### PR TITLE
Support expired List/Hash deletion

### PR DESCRIPTION
**Pull request by sherry-1001**
### What problem does this PR solve?

 Add background thread  for expired collection deletion.

### Check List 

Tests 

- Unit test
    All tests passed.
    Add `TestBackgroundDestoryCollections` test case.
- Integration test
   Bench without expired collections.
- No code

Side effects
  Not found

### CODE REVIEW

File: engine/generic_list.hpp [hunk 1]
```diff
@@ -575,16 +577,28 @@ class GenericList final : public Collection {
   }
 };
 
+template <RecordType ListType, RecordType DataType>
```
Comment (ZiyanShi) [inline engine/generic_list.hpp:580]: Put this in Collection class.

File: engine/generic_list.hpp [hunk 2]
```diff
@@ -944,6 +958,6 @@ class GenericListBuilder final {
       }
     }
   }
-};
+};  // namespace KVDK_NAMESPACE
```
Comment (ZiyanShi) [inline engine/generic_list.hpp:961]: This is the end of the class.

File: engine/kv_engine.cpp [hunk 3]
```diff
@@ -1517,6 +1532,10 @@ Status KVEngine::StringSetImpl(const StringView& key, const StringView& value,
   ExpireTimeType expired_time =
       TimeUtils::TTLToExpireTime(write_options.ttl_time, base_time);
 
+  HashEntryStatus entry_status = expired_time != kPersistTime
+                                     ? HashEntryStatus::TTL
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1536]: Rename this Status. TTL is not a status. Volatile is.

File: engine/kv_engine.cpp [hunk 4]
```diff
@@ -2032,25 +2147,32 @@ Status KVEngine::listRegisterRecovered() {
...
-  auto token = version_controller_.GetLocalSnapshotHolder();
-  list->Destroy([&](DLRecord* lrec) { delay_free(lrec, token.Timestamp()); });
+  auto ts = local ? version_controller_.GetLocalSnapshotHolder().Timestamp()
+                  : version_controller_.GetGlobalSnapshotToken()->Timestamp();
+  list->Destroy([&](DLRecord* lrec) { delay_free(lrec, ts); });
   return Status::Ok;
 }
 
 Status KVEngine::listDestroy(List* list) {
+  {
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:2167]: This one is just a wrapper. Put the actual logic in `Status KVEngine::listDestroy(List* list, DelayFree delay_free, bool local)`

File: engine/kv_engine.cpp [hunk 5]
```diff
@@ -2111,12 +2233,49 @@ Status KVEngine::listFind(StringView key, List** list, bool init_nx,
...
+      std::lock_guard<std::mutex> guard2{lists_mu_};
+      lists_.emplace(*list);
     }
     guard = (*list)->AcquireLock();
     return registerCollection(*list);
   }
 }
 
+template <typename DelayFree>
+Status KVEngine::hashListDestroy(HashList* hlist, DelayFree delay_free,
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:2245]: Wrong file.

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: engine/kv_engine.cpp [hunk 6]
```diff
@@ -92,6 +92,7 @@ void KVEngine::startBackgroundWorks() {
   bg_threads_.emplace_back(&KVEngine::backgroundOldRecordCleaner, this);
   bg_threads_.emplace_back(&KVEngine::backgroundDramCleaner, this);
   bg_threads_.emplace_back(&KVEngine::backgroundPMemUsageReporter, this);
+  bg_threads_.emplace_back(&KVEngine::backgroundDestoryCollections, this);
```
Comment (JiayuZzz) [inline engine/kv_engine.cpp:95]: typo. Destroy

File: engine/version/old_records_cleaner.cpp [hunk 7]
```diff
@@ -41,6 +41,18 @@ void OldRecordsCleaner::PushToPendingFree(void* addr, TimeStampType ts) {
   }
 }
 
+bool OldRecordsCleaner::TryGlobalPendingFree(
```
Comment (JiayuZzz) [inline engine/version/old_records_cleaner.cpp:44]: This is nothing about global. Just TryFreePendingSpace or something.

File: engine/hash_table.cpp [hunk 8]
```diff
@@ -233,7 +233,7 @@ Status HashTable::SearchForRead(const KeyHashHint& hint, const StringView& key,
 void HashTable::Insert(
     const KeyHashHint& hint, HashEntry* entry_ptr, RecordType type, void* index,
     PointerType index_type,
-    HashEntryStatus entry_status /*= HashEntryStatus::Persist*/) {
+    KeyStatus entry_status /*= KeyStatus::Persist*/) {
```
Comment (ZiyanShi) [inline engine/hash_table.cpp:236]: remove this strange comment.

File: engine/kv_engine.cpp [hunk 9]
```diff
@@ -48,7 +48,7 @@ KVEngine::~KVEngine() {
   GlobalLogger.Info("Waiting bg threads exit ... \n");
   closing_ = true;
   terminateBackgroundWorks();
-
+  releaseCollections();
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:51]: new + delete
acquire + release
deleteCollections()；

File: engine/kv_engine.cpp [hunk 10]
```diff
@@ -1552,15 +1566,15 @@ Status KVEngine::StringSetImpl(const StringView& key, const StringView& value,
...
+                      PointerType::StringRecord, entry_status);
   // Free existing record
   bool need_free =
       existing_record && ret.entry.GetRecordType() != StringDeleteRecord &&
       !ret.entry.IsExpiredStatus() /*Check if expired_key already handled by
                                        background cleaner*/
       ;
 
-  if (need_free) {
+  if (need_free && !ret.entry.IsExpiredStatus()) {
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1577]: !ret.entry.IsExpiredStatus() necessary?

File: engine/kv_engine.cpp [hunk 11]
```diff
@@ -1739,6 +1753,102 @@ void KVEngine::backgroundDramCleaner() {
...
+    if (old_records_cleaner_.TryFreePendingSpace(list_space_entries->front())) {
+      list_space_entries->pop_front();
+    }
+  }
+  return s;
+}
+
+void KVEngine::releaseCollections() {
+  for (auto list : lists_) {
+    delete list;
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1781]: Don't modify members in std::set. That's invalid.
You should get them by begin() one by one, erase it from std::set and then delete it.

File: engine/kv_engine.cpp [hunk 12]
```diff
@@ -2032,25 +2137,32 @@ Status KVEngine::listRegisterRecovered() {
...
+Status KVEngine::listDestroy(List* list, DelayFree delay_free, bool local) {
   // Currently, every list operation locks the whole list
   // and delay_free is not necessary.
   // We use delay_free here so that we can enable some lockless
   // operations in the future.
   while (list->Size() > 0) {
-    auto token = version_controller_.GetLocalSnapshotHolder();
-    list->PopFront(
-        [&](DLRecord* elem) { delay_free(elem, token.Timestamp()); });
+    auto ts = local ? version_controller_.GetLocalSnapshotHolder().Timestamp()
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:2146]: This is invalid use of a RAII object.
This token should outlive the list->PopFront() function.
Actually destroy a List does not need the token, you can use version_controller.GetTimeStamp();

File: engine/kv_engine.hpp [hunk 13]
```diff
@@ -236,18 +236,18 @@ class KVEngine : public Engine {
   // Lockless, caller should lock the key aforehand.
   // Remove key from HashTable. It's up to caller to handle the erased key
   LookupResult removeKey(StringView key) {
-    return removeImpl(key, PrimaryRecordType);
+    LookupResult result = removeImpl(key, PrimaryRecordType);
+    if (result.s == Status::Ok) {
```
Comment (ZiyanShi) [inline engine/kv_engine.hpp:240]: removeImpl() removes nothing now? How did you pass the dbtest?

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: engine/hash_table.cpp (hunk 8 above)
Comment (sherry-1001) [reply engine/hash_table.cpp:236]: it is not strange comment, it reminds developer is default parameter. Some open source project often occurs.

File: engine/kv_engine.cpp (hunk 10 above)
Comment (sherry-1001) [reply engine/kv_engine.cpp:1577]: Yes, if background thread(CleanOutDated) pushed the expired key(OldDeletedRecord) into global cleaner, but haven't purge it. The foreground thread(Set) shouldn't push it again.

File: engine/hash_table.cpp (hunk 8 above)
Comment (ZiyanShi) [reply engine/hash_table.cpp]: Default parameters are put in function declaration. Implementation does not care about that.

**Review by ZiyanShi — APPROVED** [review]

**Review by ZiyanShi — APPROVED** [review]

**Pull request merged**